ユーザー管理機能

nornia7 opened the pull request:
# What
ユーザー管理機能の実装
# Why
ユーザー登録 及び ログイン、ログアウト機能を実装するため

新規登録 # true
https://gyazo.com/bee511667d810d729ac53ccc6614e335
ログアウト # true
https://gyazo.com/74519ee3501a2964d5edaa58a9ca24b7
ログイン # true
https://gyazo.com/9cf85ba189d3f94e06bf91e73658484f

新規登録 # false
https://gyazo.com/9ec6b41f0305ac92d230490396338bf1
ログイン # false
https://gyazo.com/63b2f5f53527818376d582782f29175d

テストコード実行
https://gyazo.com/3ab2638f23d179953bdedc60397fa79f

nornia7 committed:
deviseインストール バリデーション設定 関連ViewファイルとCSSの追加

# devise
Gemfile devise 追加 インストール実行
必要カラムの追加 migration 完了
userモデル バリデーション設定 完了
アプリケーションコントローラーにuserモデル用のストロングパラメータを設定 完了

# View
views/に以下ディレクトリを追加
/devise
/shared

/layouts/application.html.erb を差し替え

assets/stylesheets/に以下ディレクトリを追加
/shared
/user

application.cssを差し替え
reset.cssの追加

nornia7 committed:
ユーザー新規登録 エラー発生時のメッセージ表示

nornia7 committed:
itemsコントローラー作成 Viewファイルの配置

TOPページを表示させるためitemコントローラーを実装
関連のviewファイルとCSSを配置しました

nornia7 committed:
ヘッダー  Viewファイル修正

・ログイン有無での表示切り替え
・ログインボタン, ログアウトボタン, 新登録ボタン 遷移設定

nornia7 committed:
新規登録時のバリデーションの追記

下記バリデーションをUserモデルへ追記

・パスワードは、半角英数字混合での入力が必須であること
・ユーザー本名は、全角（漢字・ひらがな・カタカナ）での入力が必須であること
・ユーザー本名のフリガナは、全角（カタカナ）での入力が必須であること

nornia7 committed:
rspecのインストール

nornia7 committed:
FactoryBot, Faker 導入

spec/factories/users.rb 作成
Userモデル用のインスタンスを定義

nornia7 committed:
Userモデル 単体テストコードの実装

# ユーザー新規登録
spec/models/user_spec.rb
登録できる時,登録できない時のテストーコードをそれぞれ記述

テストコード動作確認のため gem pry-rails を導入

FactoryBotのUserモデルのインスタンスの記述を一部修正

nornia7 committed:
Rubocop 実行

nornia7 committed:
ユーザーログイン ビューファイルの修正

ログインができなかった問題を修正

koyamatakumi commented on app/controllers/application_controller.rb:
```diff
@@ -1,2 +1,10 @@
+
+  def configure_permitted_parameters
+    devise_parameter_sanitizer.permit(:sign_up,
+                                      keys: %i[nickname last_name first_name last_name_kana first_name_kana birthday])
```
```
(:sign_up,keys: %i[nickname last_name first_name last_name_kana first_name_kana birthday])
```
 keys:の後に記述されている`%i`はエラーの原因になりかねないため削除しておきましょう！

koyamatakumi commented on app/controllers/items_controller.rb:
```diff
@@ -0,0 +1,3 @@
+class ItemsController < ApplicationController
+  def index; end
```
```
def index; end
```
空のindexアクションを定義されているかと存じます。
indexの後の`;`は消していただき、下記のように記述していただくと良いかと思います。
```
def index
end
```

koyamatakumi commented on app/models/user.rb:
```diff
@@ -0,0 +1,28 @@
+    validates :first_name_kana,    presence: true
+  end
+
+  validates :birthday,           presence: true
```
nicknameとbirthdayは”空では登録できない”という同じバリデーションですので
'with_options presence: true'でまとめてあげるとよりスッキリとしたコードになるかと存じます！

koyamatakumi commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,151 @@
+        @user.birthday = ''
+        @user.valid?
+        expect(@user.errors.full_messages).to include("Birthday can't be blank")
+      end
```
```
last_name
last_name_kana
first_name
first_name_kana
```
上記4つが空では登録できない場合のテストも実行しましょう！

**理由**
"空では登録できない"というバリデーションをかけているため。

koyamatakumi reviewed (commented):
レビュー依頼ありがとうございます！
コメントいたしましたので、ご確認をお願いいたします。

本日も1日頑張っていきましょう！

nornia7 committed:
pull request レビュー内容にて修正

# 修正内容

applicationコントローラー
不要記述の削除

itemsコントローラー
不要記述の削除

Userモデル
同バリデーションのカラムをwith_optionsでまとめる

Userモデルのバリデーションに関するテストコードの追加

nornia7 commented:
上記レビュー内容にて修正完了致しました。確認お願い致します。

テスト結果画面
https://gyazo.com/b8d67fe8040a6b00e4858b75c439e0ab

### 疑問点と質問
applicationコントローラーとitemsコントローラーの不要記述の削除についてですが、
コマンドラインで bundle exec rubocop -a のコマンド実行後に自動修正にて追加されたものになります。
自動修正が結果的に悪影響を及ぼすこともあるということになるんでしょうか？
また、自身で自動修正された箇所の良し悪しの判断が難しい場合は -a のオプションは付けずにチェックのみにした方がいいでしょうか？

以上よろしくお願い致します。

strongmasahiro commented on app/controllers/application_controller.rb:
```diff
@@ -1,2 +1,10 @@
+
+  def configure_permitted_parameters
+    devise_parameter_sanitizer.permit(:sign_up,
+                                      keys: [nickname last_name first_name last_name_kana first_name_kana birthday])
```
`keys: [nickname last_name first_name last_name_kana first_name_kana birthday])`
こちらの記述ですが、以下の様に修正をお願いいたします！

```
keys: [:nickname, :last_name, :first_name, :last_name_kana, :first_name_kana, :birthday])
```

strongmasahiro commented on app/models/user.rb:
```diff
@@ -0,0 +1,30 @@
+  with_options presence: true,
+               format: { with: /\A[ぁ-んァ-ヶ一-龥々ー]+\z/,
+                         message: 'is invalid. Input full-width characters' } do
+    validates :last_name,          presence: true
+    validates :first_name,         presence: true
```
```
    validates :last_name,          presence: true
    validates :first_name,         presence: true
```
こちらの`presence: true`を削除しても正しく挙動しますでしょうか？
削除しても正しく動く場合は17,18行目の`presence: true`は削除しましょう！

理由：
14行目で`with_options presence: true`で囲んでおり記述が重複してしまうためです。

strongmasahiro commented on app/models/user.rb:
```diff
@@ -0,0 +1,30 @@
+  with_options presence: true,
+               format: { with: /\A[ァ-ヶー]+\z/,
+                         message: 'is invalid. Input full-width katakana characters' } do
+    validates :last_name_kana,     presence: true
+    validates :first_name_kana,    presence: true
```
```
    validates :last_name_kana,     presence: true
    validates :first_name_kana,    presence: true
```
こちらの`presence: true`を削除しても正しく挙動しますでしょうか？
削除しても正しく動く場合は24,25行目の`presence: true`は削除しましょう！

理由：
21行目で`with_options presence: true`で囲んでおり記述が重複してしまうためです。

strongmasahiro commented on spec/factories/users.rb:
```diff
@@ -0,0 +1,14 @@
+  factory :user do
+    nickname              { Faker::Name.name }
+    email                 { Faker::Internet.free_email }
+    password = Faker::Internet.password(min_length: 6, mix_case: true)
```
`password = Faker::Internet.password(min_length: 6, mix_case: true)`
こちらの記述では、稀に「数字のみ」や「英字のみ」のパスワードが生成されてしまい、テストが失敗することが考えられます。
英数字混合のpasswordを想定した上でFakerを使用する場合は、

```
password              { '1a' + Faker::Internet.password(min_length: 6, mix_case: true) }
```
のように、確実にpasswordに英数字が含まれるように指定しましょう！

strongmasahiro commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,171 @@
+        @user.password_confirmation = '123456'
+        @user.valid?
+        expect(@user.errors.full_messages).to include('Password is invalid. Include both letters and numbers')
+      end
```
passwordが全角では登録できないことというテストも追加しましょう！

strongmasahiro reviewed (commented):
追加でコメント致しました！ご確認お願いいたします！
また、ご質問いただきありがとうございます！以下で回答させていただきました。

> 自動修正が結果的に悪影響を及ぼすこともあるということになるんでしょうか？
 また、自身で自動修正された箇所の良し悪しの判断が難しい場合は -a のオプションは付けずにチェックのみにした方がいいでしょうか？

ありがとうございます！自動修正に関しましては、稀に悪影響を及ぼすことがあります。そのため、おっしゃっていただいている様に、良し悪しの判断が難しい場合はチェックのみにしていただければと思います。

もし、ご不明点などある場合は、通話にてご質問いただければ幸いです。
何卒よろしくお願いいたします。

nornia7 commented:
レビューとRubocopに対する回答ありがとうございます。

再修正完了致しました。
重複していましたバリデーションの記述削除後もテストコードにて動作に問題がないことを確認しました。
テストコード実行結果
https://gyazo.com/38f26af39f6ca0309effe6f5a7494566

再度確認をお願い致します

nornia7 committed:
レビュー内容にて修正 #2

div-reviewer reviewed (commented):
学習お疲れ様です！お忙しい中修正と再レビュー依頼ありがとうございました。

>レビューとRubocopに対する回答ありがとうございます。
再修正完了致しました。
重複していましたバリデーションの記述削除後もテストコードにて動作に問題がないことを確認しました。

ご丁寧に返信ありがとうございます。
前回分のコメントについて、綺麗に修正できていますね！
確認も丁寧にして頂き、また最新のテスト画像も添付頂きありがとうございました。

こちらユーザー管理機能にほしい情報が全て実装できているかと思われます。
よって`LGTM`とさせていただきます！実装お疲れさまでした。
次のブランチを切る前にMergeとPullを忘れないようお願いします！

次は「商品出品機能」かと思われます！引き続き頑張りましょう！応援しております！
